controllers/nodemaintenance_controller.go

```diff
@@ -91,143 +101,132 @@ type NodeMaintenanceReconciler struct {
 func (r *NodeMaintenanceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	r.logger = log.FromContext(ctx)
 	r.logger.Info("Reconciling NodeMaintenance")
+	defer r.logger.Info("Reconcile completed")
+	emptyResult := ctrl.Result{}
 
-	// Fetch the NodeMaintenance instance
-	instance := &nodemaintenancev1beta1.NodeMaintenance{}
-	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	// Fetch the NodeMaintenance instance (nm)
+	nm := &v1beta1.NodeMaintenance{}
+	err := r.Client.Get(ctx, req.NamespacedName, nm)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			// Request object not found, could have been deleted after reconcile request.
 			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
 			// Return and don't requeue
 			r.logger.Info("NodeMaintenance not found", "name", req.NamespacedName)
-			return reconcile.Result{}, nil
+			return emptyResult, nil
 		}
 		// Error reading the object - requeue the request.
 		r.logger.Info("Error reading the request object, requeuing.")
-		return reconcile.Result{}, err
+		return emptyResult, err
 	}
 
-	// Add finalizer when object is created
-	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
-		if !ContainsString(instance.ObjectMeta.Finalizers, nodemaintenancev1beta1.NodeMaintenanceFinalizer) {
-			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, nodemaintenancev1beta1.NodeMaintenanceFinalizer)
-			if err := r.Client.Update(context.TODO(), instance); err != nil {
-				return r.onReconcileError(instance, err)
-			}
+	if err := createDrainer(r); err != nil {
+		return emptyResult, err
+	}
+	// set drain Helper context
+	r.drainer.Ctx = ctx
```

Comment:
We mentioned concurrency, right? So yes, use the context of this reconcile call. But where do you store it?

Reply:
It is stored in a newly created (and initialized from _createDrainer_) drain.Helper struct.
We will create this struct on every reconcile run, set it for *NodeMaintenanceReconciler object, and initialize its fields in _createDrainer_. Afterward, when there is no error we will set its context field (`r.drainer.Ctx`) using the context of this reconcile call. Do you suggest that check whether the struct exists?

But I assume it doesn't what you mean or hint.

Reply:
> set it for *NodeMaintenanceReconciler object

....?

Reply:
> > set it for *NodeMaintenanceReconciler object
> 
> ....?

The r *NodeMaintenanceReconciler object has a drainer field that would be initialized with a newly created struct on every reconcile run. After initialization in _createDrainer_ the context would be set as well in the _Reconcile_ function.

Reply:
So you have ONE reconciler, which is used for all reconcile calls.
And on EACH reconcile call you create a NEW drainer, which you store on that ONE reconciler.
And now reconcile is called concurrently, what will happen...?